### ISSUE TITLE
SCF Convergence and Integral Screening

### ISSUE THREAD

**Issue by zachglick** [psi4/psi4#2114]
The following SCF calculation fails to converge when run with psi4/master, but it completed okay with psi4/v1.3.2. This is due to the default integral screening threshold (`INTS_TOLERANCE`) changing from 0.0 to 1e-12 for the SCF module. The calculation converges fine when `INTS_TOLERANCE` is reset to 0.0:

```
memory 64 GB

molecule mol {
  0 1 
  O -1.43902644 2.20001916 0.52666667
  C -0.23262529 0.21202727 0.63286928
  N -1.24873102 -2.11271381 0.47828968
  C -3.7777515 -2.85877768 0.08730535
  C -4.46145441 -5.34849185 0.74152853
  C -2.56341349 -7.19739989 1.87800982
  C -6.94682221 -6.16901094 0.3643392
  C -8.75094374 -4.55140537 -0.66556154
  C -8.08273658 -2.09627319 -1.33509151
  C -5.61324248 -1.25402226 -0.96753978
  C 2.6282311 0.28062433 1.05654588
  C 3.98467651 2.83553405 0.63381414
  N 6.54374363 2.70155247 -0.40515728
  C 8.61979675 2.25141971 0.96961847
  O 8.56480572 1.89615119 3.26884825
  H -0.04232987 -3.54814977 0.82014114
  H -3.42815216 -8.99585225 2.39447197
  H -1.03934937 -7.57742382 0.54424112
  H -1.71851694 -6.40333698 3.5817869
  H -7.49559868 -8.07309898 0.86738429
  H -10.66693706 -5.20506164 -0.95166608
  H -9.47697652 -0.84262888 -2.14880758
  H -5.16216486 0.65044373 -1.5435283
  H 3.24390387 -0.55898099 2.83213254
  H 3.51375676 -1.09698602 -0.19445282
  H 3.04888413 3.57366108 -1.05446718
  H 3.38827894 4.41440023 1.82490852
  H 6.75331425 3.01713673 -2.28241121
  H 10.34700643 2.24329388 -0.20333453
  symmetry c1
  no_com
  no_reorient
  units bohr
}

# set ints_tolerance 0.0 # uncomment to fix

energy('pbe0/aug-cc-pv_tpd_z')
```
~~[screen_zero.out.txt](https://github.com/psi4/psi4/files/5984125/screen_zero.out.txt)~~
~~[screen_default.out.txt](https://github.com/psi4/psi4/files/5984126/screen_default.out.txt)~~
[failed_jobs.zip](https://github.com/psi4/psi4/files/5985739/failed_jobs.zip)

I can post additional similar systems that fail at the same level of theory.

Does this warrant changing the `INTS_TOLERANCE` default to a more conservative value? Or is this something that would be fixed by a more robust SCF solver?

**Comment by JonathonMisiewicz** [comment]
Look at iteration 1: we're seeing a change in the iteration energy of over a mH due to this integral screening. This strongly suggests a bona fide bug. If you have additional systems that fail, this seems like a bug that is going to afflict a _lot_ of our users. I highly suggest we fast-track this for investigation but I would rather avoid being the one to debug this directly. (I'm trying to get into paper-writing mode.)

**Comment by hokru** [comment]
Please post the other systems as well.

For a normal organic molecule one should not need 0.0 tolerance. 
Looking at the `screen_zero.out.txt` output I see negative virtual orbital energies in the converged SCF, which tells me something is still wrong. Zwitterionic system?

**Comment by susilehtola** [comment]
I assume this is commit 4ad38a60ec?

A cutoff of `1e-12` should not cause any issues. Even `1e-10` is fine for most systems.

**Comment by JonathonMisiewicz** [comment]
Susi raises a great point. It looks like you're on a local branch with at least some commits that haven't been published to Psi. `Git: Rev {localsapt} 5063628 dirty`

**Comment by zachglick** [comment]
Okay, I've updated the body of this issue with five SCF outputs that fail to converge with the default `INTS_TOLERANCE` (1e-12), but do converge when `INTS_TOLERANCE` is explicitly set to 0.0. All calculations were done with a clean version of psi4 master (the previous outputs had some commits from an unrelated project).

@susilehtola yes, that is the relevant commit in which the default `INTS_TOLERANCE` changed. I agree that the new default appears conservative enough, and I'm surprised such a small threshold has this kind of effect, even for a large, diffuse basis set. Do you know of any integral screening protocols specific to diffuse functions?

@hokru your observation that some virtual orbitals have negative energies is true for all zero-screening outputs. Would this imply that the SCF solution is not the global minimum? I don't have any specifics about these structures, they're part of a large dataset for a ML project. However, they all look relatively ordinary.

@JonathonMisiewicz Your observation that `INTS_TOLERANCE` has a large effect on the very first SCF iteration also is true of all five outputs. In the second example, the first SCF iteration differs between the two outputs by almost four(!) hartree.

**Comment by susilehtola** [comment]
btw are you only running DFT? The Karlsruhe or Jensen basis sets are much more accurate / cheaper than the Dunning ones..

**Comment by susilehtola** [comment]
Anyway, the logical place to start debugging is SAD and the first iteration after it.

|System|Difference in SAD energy|Difference in energy of first iteration|
|----------|----------------------------------|-----------------------------------------------|
|1| 2.387e-12|-1.405e-03|
|2| 6.253e-12|-3.704e+00|
|3| 1.364e-11|-5.389e-01|
|4| 8.868e-12|-1.833e+00|
|5|-8.868e-12|-3.321e+00|

The SAD energies are an exact match, so one would assume that the SAD densities also agree exactly. Something is causing the energy of the first iteration to change by quite a lot. This shouldn't happen: if the SAD density matrices are the same, the Fock matrices built from the SAD density matrices should also match; the orbitals in the first iteration should be practically the same. One can't compare after the first iteration, since the orbitals and densities will not be the same.

So, here are the possibilities for disagreeing results.

1. Atoms in SAD converge to different solutions
2. SAD Fock matrix is not computed properly (same input density leads to differences between runs)
3. First Fock matrix is not computed properly (same input density leads to differences between runs)

1 and 2 are unlikely, since the SAD energies agree.

**Comment by zachglick** [comment]
I am only doing DFT for now Susi, thanks for the tip about those basis sets. 

I've determined that the bug was definitely introduced in the libint2 PR (#1721). Prior to that PR, `INTS_TOLERANCE` worked as expected: a value of 1e-12 led to a minor speed up and a negligible change in the energy. I further narrowed the problem down to the `engine_precision` parameter passed to the libint2 engine:

https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/eri.cc#L182

If I instead set `engine_precision` to `max_precision`, the SCF calculation proceeds normally. This suggests that the current value of `engine_precision` is too loose. Does anyone know more about what this parameter is, and where the current formula comes from? I assume the `-4` is for the four indexes of an ERI.

**Comment by susilehtola** [comment]
Looking at the libint header file at https://github.com/evaleev/libint/blob/de668f24fafe6fc28cbb52f6fa4738a9912756f1/include/libint2/engine.h#L477 I see that

> precision specifies the target precision with which the integrals will be computed; the default is the "epsilon" of scalar_type type, given by std::numeric_limits<scalar_type>::epsilon(). Currently precision control is implemented for two-body integrals only. The precision control is somewhat empirical, hence be conservative.

Sounds to me like that parameter should not be changed, especially since it leads to the integrals being computed in an inaccurate manner. What `INTS_TOLERANCE` should do is just control the screening threshold for those integrals that do get computed (the others are ignored altogether), not affect the precision at which the non-zero integrals are computed.

**Comment by zachglick** [comment]
Good find, there's some additional info in that header file that explains how the precision parameter is used to screen primitives.

If we use the default libint2 engine precision parameter (`std::numeric_limits<double>::epsilon()`) we would still have a problem, since the current engine precision parameter (`std::numeric_limits<double>::epsilon() * std::pow((double)max_nprim, -4.0) * 1.e-2`) is smaller/tighter and causes convergence issues. I think we need something closer to 0.0.

I've found that tightening the parameter by a factor of 10 (setting it to `std::numeric_limits<double>::epsilon() * std::pow((double)max_nprim, -4.0) * 1.e-3`) fixes all of the errors I've encountered. Any objections to this? Also, does anyone know how this parameter was chosen in the first place?

**Comment by susilehtola** [comment]
My guess is it was just chosen based on the name of the function, without deeper thought into its documentation. After all, "engine precision" in QM talk usually means the threshold at which you throw everything out, and calculate the rest at full double precision.

**Comment by susilehtola** [comment]
Is there a performance handicap in using `0.0`?

**Comment by zachglick** [comment]
I ran a HF/aTZ calculation on decane with `scf_type direct` and `df_scf_guess false` (meaning the majority of the calculation is spent computing integrals). Here's the walltime (in seconds) of the JK step for different values of `engine_precision`:

| 0.0 <br /> (no screening) | ε * (max_nprim ^ -4.0) * 1.e-3  <br /> (proposed change) | ε * (max_nprim ^ -4.0) * 1.e-2 <br /> (current value) | ε <br /> (libint2 default) |
|:------------------:|-----------------------------------------------------------|------------------------------------------------------|-----------------------------|
|        3107        | 3031                                                       | 3041                                                 | 2843                        |

Looks like there'd be a small but non-negligible performance hit if we used 0.0. The current precision setting already performs close to 0.0 (relative to the libint2 default). There's basically no change in performance if we tighten the current setting by a factor of 10.

This conventional HF example is also a kind of worst-case scenario, since integral calculation is not the rate-limiting step in most Psi4 calculations (like anything with DF). I think it wouldn't hurt to tighten (or completely turn off) primitive screening, although there could be other systems or basis sets for which primitive screening leads to larger performance gains.

**Comment by TiborGY** [comment]
> 
> This conventional HF example is also a kind of worst-case scenario, since integral calculation is not the rate-limiting step in most Psi4 calculations (like anything with DF). I think it wouldn't hurt to tighten (or completely turn off) primitive screening, although there could be other systems or basis sets for which primitive screening leads to larger performance gains.

Maybe the ANO basis set family would suffer from this? While I do not use them, IIRC they are notorious for having an extremely high number of primitives, and for slowing many integral packages to a glacial pace.

**Comment by susilehtola** [comment]
> Maybe the ANO basis set family would suffer from this? While I do not use them, IIRC they are notorious for having an extremely high number of primitives, and for slowing many integral packages to a glacial pace.

You don't even need to go to ANO; just take the cc sets for transition metals. I ran a simple benchmark [here](https://mattermodeling.stackexchange.com/questions/3794/which-software-is-good-with-generally-contracted-basis-sets) for Zn2 with the cc-pVDZ basis set. Zn and Zn2 ran in 19 and 216 seconds of wall time with Psi4, compared to 0.7 and 1.4 seconds with PySCF, respectively.

**Comment by zachglick** [comment]
Fixed by #2139. In the future, it may be worthwhile to reinvestigate primitive screening in libint2.
